go/vt/vttablet/tabletserver/repltracker/writer.go

```diff
@@ -133,34 +163,52 @@ func (w *heartbeatWriter) Open() {
 	// block this thread, and we could end up in a deadlock.
 	// Instead, we try creating the database and table in each tick which runs in a go routine
 	// keeping us safe from hanging the main thread.
-	w.appPool.Open(w.env.Config().DB.AppWithDB())
-	w.allPrivsPool.Open(w.env.Config().DB.AllPrivsWithDB())
-	if w.onDemandDuration == 0 {
-		w.enableWrites(true)
-		// when onDemandDuration > 0 we only enable writes per request
-	} else {
+
+	if w.keyspaceShard != testKeyspaceShard {
+		// testKeyspaceShard indicates we're running in a unit test, in which case mock
+		// pools will have been opened.
+		w.appPool.Open(w.env.Config().DB.AppWithDB())
+		w.allPrivsPool.Open(w.env.Config().DB.AllPrivsWithDB())
+	}
+
+	if w.onDemandDuration > 0 {
+		// Clients are given access to RequestHeartbeats(). But such clients can also abuse
+		// the system by initiating heartbeats too frequently. We rate-limit these requests.
+		rateLimiter := timer.NewRateLimiter(w.onDemandDuration / 4)
+		atomic.StorePointer(w.unsafeOnDemandRequestsRateLimiter(), unsafe.Pointer(rateLimiter))
+	}
+
+	switch w.configType {
+	case HeartbeatConfigTypeNone:
+		// Do not kick any heartbeats
+		return
+	case HeartbeatConfigTypeAlways:
+		// Heartbeats are always on
+		w.enableWrites()
+	case HeartbeatConfigTypeOnDemand:
 		// A one-time kick off of heartbeats upon Open()
 		go w.RequestHeartbeats()
 	}
-
-	w.isOpen = true
 }
 
 // Close closes the heartbeatWriter's db connection and stops the periodic ticker.
 func (w *heartbeatWriter) Close() {
-	if !w.enabled {
-		return
-	}
 	w.mu.Lock()
 	defer w.mu.Unlock()
 	if !w.isOpen {
 		return
 	}
+	w.isOpen = false
```

Comment:
IMO this should remain just before:
```
log.Info("Heartbeat Writer: closed")
```

Reply:
Now using `defer` to ensure the value is changed.